lib/kafkat/command/drain.rb

```diff
@@ -0,0 +1,83 @@
+module Kafkat
+  module Command
+    class Drain < Base
+      register_as 'drain'
+
+      usage 'drain <broker id> [topic] [--brokers <ids>]',
+            'Reassign partitions from a specific broker to destination brokers.'
+
+      # For each partition (of speicified topic) on the source broker, the command is to
+      # assign the partition to one of the destination brokers that does not already have
+      # this partition, along with existing brokers to achieve minimal movement of data.
+      # To help distribute data evenly, if there are more than one destination brokers
+      # meet the requirement, the command will always choose the brokers with the lowest
+      # number of partitions of the involving topic.
+      #
+      # In order to find out the broker with lowest number of partitions, the command maintain
+      # a hash table with broker id as key and number of partitions as value. The hash table
+      # will be updated along with assignment.
+      def run
+        source_broker_id = ARGV[0] && ARGV.shift.to_i
+        if source_broker_id.nil?
+          puts "You must specify a broker ID."
+          exit 1
+        end
+
+        topic_name = ARGV.shift unless ARGV[0] && ARGV[0].start_with?('--')
+        topics = topic_name && zookeeper.get_topics([topic_name])
+        topics ||= zookeeper.get_topics
+
+        opts = Trollop.options do
+          opt :brokers, "destination broker IDs", type: :string
+        end
+
+        destination_broker_ids = opts[:brokers] && opts[:brokers].split(',').map(&:to_i)
+        destination_broker_ids ||= zookeeper.get_brokers.values.map(&:id)
+        destination_broker_ids.delete(source_broker_id)
+        all_brokers_id = zookeeper.get_brokers.values.map(&:id)
+
+        unless (inactive_broker_ids = destination_broker_ids - all_brokers_id).empty?
+          print "ERROR: Broker #{inactive_broker_ids} are not currently active.\n"
+          exit 1
+        end
+
+        assignments = generate_assignments(source_broker_id, topics, destination_broker_ids)
+        prompt_and_execute_assignments(assignments)
+      end
+
+      def generate_assignments(source_broker_id, topics, destination_broker_ids)
+        assignments = []
+
+        topics.each do |_, t|
+          num_partitions_on_broker = Hash.new{0}
```

Comment:
Nice logic here. This will nicely balance the partitions.
Could you extract that in a method, or at least document the intent.